language/move-prover/stackless-bytecode-generator/src/function_target.rs

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    annotations::Annotations,
+    stackless_bytecode::{AttrId, Bytecode},
+};
+use spec_lang::{
+    ast::Condition,
+    env::{FunId, FunctionEnv, Loc, TypeParameter},
+    symbol::{Symbol, SymbolPool},
+    ty::Type,
+};
+use std::collections::BTreeMap;
+use vm::file_format::CodeOffset;
+
+/// A FunctionTarget is a drop-in replacement for a FunctionEnv which allows to rewrite
+/// and analyze bytecode and parameter/local types. It encapsulates a FunctionEnv and information
+/// which can be rewritten using the `FunctionTargetsHolder` data structure.
+#[derive(Debug)]
+pub struct FunctionTarget<'env> {
+    pub func_env: &'env FunctionEnv<'env>,
+    pub data: &'env FunctionTargetData,
+}
+
+/// Holds the owned data belonging to a FunctionTarget, which can be rewritten using
+/// the `FunctionTargetsHolder::rewrite` method.
+#[derive(Debug)]
+pub struct FunctionTargetData {
+    pub code: Vec<Bytecode>,
+    pub local_types: Vec<Type>,
+    pub return_types: Vec<Type>,
+    pub locations: BTreeMap<AttrId, Loc>,
+    pub annotations: Annotations,
+}
+
+impl<'env> FunctionTarget<'env> {
+    /// Returns the name of this function.
+    pub fn get_name(&self) -> Symbol {
+        self.func_env.get_name()
+    }
+
+    /// Gets the id of this function.
+    pub fn get_id(&self) -> FunId {
+        self.func_env.get_id()
+    }
+
+    /// Shortcut for accessing the symbol pool.
+    pub fn symbol_pool(&self) -> &SymbolPool {
+        self.func_env.module_env.symbol_pool()
+    }
+
+    /// Returns the location of this function.
+    pub fn get_loc(&self) -> Loc {
+        self.func_env.get_loc()
+    }
+
+    /// Returns the location of the bytecode at the given offset.
```

Comment:
Nit: might be useful for the comment to specify when the behavior of the function can change after mutation (like  `get_byecode_loc`, but unlike `get_id`)? It could also be part of the naming convention (e.g., `get_bytecode_loc_unstable`)